src/annotator/test/guest-test.js

```diff
@@ -420,15 +420,18 @@ describe('Guest', () => {
     });
 
     describe('on "deleteAnnotation" event', () => {
-      it('detaches annotation', () => {
+      it('prevents saving of annotations that are currently anchoring', async () => {
         const guest = createGuest();
-        const callback = sinon.stub();
-        guest._emitter.subscribe('anchorsChanged', callback);
+        const ann = { target: [], uri: 'uri', $tag: 'tag1' };
 
-        emitSidebarEvent('deleteAnnotation', 'tag1');
+        emitSidebarEvent('loadAnnotations', [ann]);
+        emitSidebarEvent('deleteAnnotation', ann.$tag);
```

Comment:
If you invert both lines the test should not pass.

I think there is a caveat to this solution: if the deletion of the annotation happens _before_ the loading the annotation will be anchored. This solution works if the deletion happens _while_ the anchoring is taking place.

Reply:
> Depending of how the browser decides to execute the waiting anchoring tasks the race condition could still persist.

The code looks safe at present. `this.anchor` is called synchronously for all annotations in the `loadAnnotations` event handler in `Guest`. This event handler is in turn called in response to our code (PortRPC) receiving a message via a MessagePort. The MessagePort in turn will deliver events in the other they were posted by the sending frame. As long as the sidebar calls `loadAnnotations` before `deleteAnnotation`, everything should work.

There is a refactoring hazard here though. If someone where to add async steps in the `loadAnnotations` event handler or `anchor` method before `this._annotations.add` were called, then it could introduce a race. This can be prevented by a test like this which calls `loadAnnotations` followed by `deleteAnnotation` as fast as possible. In the actual app, the fastest a `deleteAnnotation` call can follow a `loadAnnotations` call is on the next macrotask. Here it is being done in the same event loop turn.

Reply:
> The code looks safe to me at present. this.anchor is called synchronously for all annotations in the loadAnnotations event handler in Guest. This event handler is in turn called in response to our code (PortRPC) receiving a message via a MessagePort. The MessagePort in turn will deliver events in the other they were posted by the sending frame. As long as the sidebar calls loadAnnotations before deleteAnnotation, everything should work.

Thanks for the explanation. I agree everything should work fine to avoid the race condition.

Reply:
> There is a refactoring hazard here though. If someone where to add async steps in the loadAnnotations event handler or anchor method before this._annotations.add were called, then it could introduce a race. This can be prevented by a test like this which calls loadAnnotations followed by deleteAnnotation as fast as possible. In the actual app, the fastest a deleteAnnotation call can follow a loadAnnotations call is on the next macrotask. Here it is being done in the same event loop turn.

I agree, I will try to work on that test. In the past I found challenging to simulate race conditions. Let's see...